plan/new_physical_plan_builder.go

```diff
@@ -101,172 +101,185 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
+		if includeTableScan && len(innerJoinKeys) == 1 {
+			pkCol := x.getPKIsHandleCol()
+			if pkCol != nil && innerJoinKeys[0].Equal(pkCol, nil) {
+				return p.constructIndexJoin(innerJoinKeys, outerJoinKeys, outerIdx)
 			}
-			if useTableScan {
-				innerTask, err = x.convertToTableScan(&requiredProp{taskTp: rootTaskType})
-				if err != nil {
-					return nil, errors.Trace(err)
-				}
-				break
-			}
-			for _, indexInfo := range indices {
-				matchedOffsets := joinKeysMatchIndex(innerJoinKeys, indexInfo)
-				if matchedOffsets == nil {
-					continue
-				}
-				usedIndexInfo = indexInfo
-				newOuterJoinKeys := make([]*expression.Column, len(outerJoinKeys))
-				newInnerJoinKeys := make([]*expression.Column, len(innerJoinKeys))
-				for i, offset := range matchedOffsets {
-					newOuterJoinKeys[i] = outerJoinKeys[offset]
-					newInnerJoinKeys[i] = innerJoinKeys[offset]
-				}
-				outerJoinKeys = newOuterJoinKeys
-				innerJoinKeys = newInnerJoinKeys
-				break
+		}
+		for _, indexInfo := range indices {
+			matchedOffsets := joinKeysMatchIndex(innerJoinKeys, indexInfo)
+			if matchedOffsets == nil {
+				continue
 			}
-			if usedIndexInfo != nil {
-				innerTask, err = x.convertToIndexScan(&requiredProp{taskTp: rootTaskType}, usedIndexInfo)
-				if err != nil {
-					return nil, errors.Trace(err)
-				}
-				break
+			usedIndexInfo = indexInfo
+			newOuterJoinKeys := make([]*expression.Column, len(outerJoinKeys))
+			newInnerJoinKeys := make([]*expression.Column, len(innerJoinKeys))
+			for i, offset := range matchedOffsets {
+				newOuterJoinKeys[i] = outerJoinKeys[offset]
+				newInnerJoinKeys[i] = innerJoinKeys[offset]
 			}
-			return nil, nil
-		case *Selection:
-			rightConds = append(rightConds, x.Conditions...)
-			innerChild = innerChild.Children()[0].(LogicalPlan)
-		default:
-			return nil, nil
-		}
-		if innerTask != nil {
+			outerJoinKeys = newOuterJoinKeys
+			innerJoinKeys = newInnerJoinKeys
 			break
 		}
+		if usedIndexInfo != nil {
+			return p.constructIndexJoin(innerJoinKeys, outerJoinKeys, outerIdx)
+		}
 	}
-	join := PhysicalIndexJoin{
-		LeftConditions:  leftConds,
-		RightConditions: rightConds,
-		OtherConditions: p.OtherConditions,
-		Outer:           p.JoinType != InnerJoin,
-		OuterJoinKeys:   outerJoinKeys,
-		InnerJoinKeys:   innerJoinKeys,
-		DefaultValues:   p.DefaultValues,
-	}.init(p.allocator, p.ctx, p.children[outerIdx], p.children[1-outerIdx])
-	task := join.attach2Task(outerTask, innerTask)
-	task = prop.enforceProperty(task, p.ctx, p.allocator)
-	return task, nil
+	return nil
+}
+
+func (p *PhysicalIndexJoin) getChildrenPossibleProps(prop *requiredProp) [][]*requiredProp {
+	if !prop.isEmpty() {
+		return nil
+	}
+	requiredProps1 := make([]*requiredProp, 2)
+	requiredProps1[p.outerIndex] = &requiredProp{taskTp: rootTaskType}
+	requiredProps1[1-p.outerIndex] = &requiredProp{taskTp: copSingleReadTaskType, cols: p.InnerJoinKeys}
```

Comment:
If it cannot do single read, it will generate a invalid task that never be choosed.